### PR TITLE
feat: add tlog SET verification function

### PR DESCRIPTION
**Pull request by asraa**
Signed-off-by: Asra Ali <asraa@google.com>

* feat: adds tlog SET verification function against TransparencyLogEntry (defined in protobuf-specs)

Part of https://github.com/sigstore/sigstore-go/issues/11

### CODE REVIEW

**Review by cpanato — DISMISSED** [review]
Cool

**Comment by asraa** [comment]
Just fixed lint!

**Review by cpanato — DISMISSED** [review]

**Comment by asraa** [comment]
@haydentherapper want to do a quick skim? this is mostly stolen from rekor's verification/cosign's verification except that it works on the protobuf log entry.

File: pkg/tlog/common.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/hex"
+	"errors"
+
+	rekor_v1 "github.com/sigstore/protobuf-specs/gen/pb-go/rekor/v1"
+)
+
+// ComputeLogID generates a SHA256 hash of a DER-encoded public key.
+func ComputeLogID(pub crypto.PublicKey) (string, error) {
```
Comment (Hayden-IO) [inline pkg/tlog/common.go:31]: Can you add tests for these? (gotta get that 100% codecov :) )

File: pkg/tlog/common.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	pubBytes, err := x509.MarshalPKIXPublicKey(pub)
+	if err != nil {
+		return "", err
+	}
+	digest := sha256.Sum256(pubBytes)
+	return hex.EncodeToString(digest[:]), nil
+}
+
+func GetLogID(entry *rekor_v1.TransparencyLogEntry) (string, error) {
```
Comment (Hayden-IO) [inline pkg/tlog/common.go:41]: nit, function comment for exported function

File: pkg/tlog/common.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	"crypto"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/hex"
+	"errors"
+
+	rekor_v1 "github.com/sigstore/protobuf-specs/gen/pb-go/rekor/v1"
+)
+
+// ComputeLogID generates a SHA256 hash of a DER-encoded public key.
```
Comment (Hayden-IO) [inline pkg/tlog/common.go]:
```suggestion
// ComputeLogID generates a SHA256 hash of a DER-encoded public key, which is
// the log ID of a transparency log.
```

File: pkg/tlog/verify.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	"errors"
+	"fmt"
+
+	"github.com/cyberphone/json-canonicalization/go/src/webpki.org/jsoncanonicalizer"
+	rekor_v1 "github.com/sigstore/protobuf-specs/gen/pb-go/rekor/v1"
+	"github.com/sigstore/sigstore/pkg/signature"
+	"github.com/sigstore/sigstore/pkg/signature/options"
+)
+
+type VerificationPayload struct {
```
Comment (Hayden-IO) [inline pkg/tlog/verify.go:36]: nit, comment

File: pkg/tlog/verify.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	Body           interface{} `json:"body"`
+	IntegratedTime int64       `json:"integratedTime"`
+	LogIndex       int64       `json:"logIndex"`
+	LogID          string      `json:"logID"`
+}
+
+// VerifyTlogSET verifies the SET for the given TransparencyLogEntry using the trusted
```
Comment (Hayden-IO) [inline pkg/tlog/verify.go]:
```suggestion
// VerifyTlogSET verifies the Signed Entry Timestamp (SET) for the given TransparencyLogEntry using the trusted
```

File: pkg/tlog/verify.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	if err != nil {
+		return fmt.Errorf("getting entry log ID: %w", err)
+	}
+	verifier, ok := trustedKeys[entryLogID]
+	if !ok {
+		return errors.New("rekor log public key not found for payload")
+	}
+
+	// Extract the SET from the tlog entry
+	if entry.GetInclusionPromise() == nil {
```
Comment (Hayden-IO) [inline pkg/tlog/verify.go:76]: this field makes me happy, excellent naming!

File: pkg/tlog/verify_test.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	"testing"
+
+	"github.com/cyberphone/json-canonicalization/go/src/webpki.org/jsoncanonicalizer"
+	common_v1 "github.com/sigstore/protobuf-specs/gen/pb-go/common/v1"
+	rekor_v1 "github.com/sigstore/protobuf-specs/gen/pb-go/rekor/v1"
+	"github.com/sigstore/sigstore/pkg/signature"
+)
+
+const (
+	encodedBody = "ayJhcGlWZXJzaW9uIjoiMC4wLjEiLCJraW5kIjoicmVrb3JkIiwic3BlYyI6eyJkYXRhIjp7Imhhc2giOnsiYWxnb3JpdGhtIjoic2hhMjU2IiwidmFsdWUiOiJlY2RjNTUzNmY3M2JkYWU4ODE2ZjBlYTQwNzI2ZWY1ZTliODEwZDkxNDQ5MzA3NTkwM2JiOTA2MjNkOTdiMWQ4In19LCJzaWduYXR1cmUiOnsiY29udGVudCI6Ik1FWUNJUUQvUGRQUW1LV0MxKzBCTkVkNWdLdlFHcjF4eGwzaWVVZmZ2M2prMXp6Skt3SWhBTEJqM3hmQXlXeGx6NGpwb0lFSVYxVWZLOXZua1VVT1NvZVp4QlpQSEtQQyIsImZvcm1hdCI6Ing1MDkiLCJwdWJsaWNLZXkiOnsiY29udGVudCI6IkxTMHRMUzFDUlVkSlRpQlFWVUpNU1VNZ1MwVlpMUzB0TFMwS1RVWnJkMFYzV1VoTGIxcEplbW93UTBGUldVbExiMXBKZW1vd1JFRlJZMFJSWjBGRlRVOWpWR1pTUWxNNWFtbFlUVGd4UmxvNFoyMHZNU3R2YldWTmR3cHRiaTh6TkRjdk5UVTJaeTlzY21sVE56SjFUV2haT1V4alZDczFWVW8yWmtkQ1oyeHlOVm80VERCS1RsTjFZWE41WldRNVQzUmhVblozUFQwS0xTMHRMUzFGVGtRZ1VGVkNURWxESUV0RldTMHRMUzB0Q2c9PSJ9fX19"
```
Comment (Hayden-IO) [inline pkg/tlog/verify_test.go]: Is this correct, this decodes to this for me:

```
k"apiVersion":"0.0.1","kind":"rekord","spec":{"data":{"hash":{"algorithm":"sha256","value":"ecdc5536f73bdae8816f0ea40726ef5e9b810d914493075903bb90623d97b1d8"}},"signature":{"content":"MEYCIQD/PdPQmKWC1+0BNEd5gKvQGr1xxl3ieUffv3jk1zzJKwIhALBj3xfAyWxlz4jpoIEIV1UfK9vnkUUOSoeZxBZPHKPC","format":"x509","publicKey":{"content":"LS0tLS1CRUdJTiBQVUJMSUMgS0VZLS0tLS0KTUZrd0V3WUhLb1pJemowQ0FRWUlLb1pJemowREFRY0RRZ0FFTU9jVGZSQlM5amlYTTgxRlo4Z20vMStvbWVNdwptbi8zNDcvNTU2Zy9scmlTNzJ1TWhZOUxjVCs1VUo2ZkdCZ2xyNVo4TDBKTlN1YXN5ZWQ5T3RhUnZ3PT0KLS0tLS1FTkQgUFVCTElDIEtFWS0tLS0tCg=="}}}}
```

note the `k`

File: pkg/tlog/verify_test.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+			entry: tlogEntry,
+		},
+		{
+			name:    "fail: missing trusted tlog key",
+			entry:   tlogEntry,
+			keys:    map[string]signature.Verifier{},
+			wantErr: true,
+		},
+		{
+			name: "fail: missing  entry log id",
```
Comment (Hayden-IO) [inline pkg/tlog/verify_test.go]:
```suggestion
			name: "fail: missing entry log id",
```

**Review by Hayden-IO — COMMENTED** [review]
mostly just some nits

File: pkg/tlog/verify_test.go (hunk 7 above)
Comment (asraa) [reply pkg/tlog/verify_test.go]: oh, it didn't need to be, for clarity i'll just put in an explicit "dummy" value. SET verification doesn't do any unmarshalling of this field.

File: pkg/tlog/verify.go (hunk 6 above)
Comment (asraa) [reply pkg/tlog/verify.go:76]: it's protobuf generated code!

**Comment by asraa** [comment]
Updated with more comments and tests

**Review by Hayden-IO — APPROVED** [review]

**Pull request merged**